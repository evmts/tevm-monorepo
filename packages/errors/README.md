<p align="center">
  @tevm/errors
<p>

[![CI](https://github.com/evmts/tevm-monorepo/actions/workflows/e2e.yml/badge.svg)](https://github.com/evmts/tevm-monorepo/actions/workflows/e2e.yml)
[![CI](https://github.com/evmts/tevm-monorepo/actions/workflows/unit.yml/badge.svg)](https://github.com/evmts/tevm-monorepo/actions/workflows/unit.yml)
<a href="https://www.npmjs.com/package/@tevm/errors" target="\_parent">
<img alt="" src="https://img.shields.io/npm/dm/@tevm/errors.svg" />
</a>
<a href="https://bundlephobia.com/package/@tevm/errors@latest" target="\_parent">
<img alt="" src="https://badgen.net/bundlephobia/minzip/@tevm/errors" />
</a>

# @tevm/errors

`@tevm/errors` is the official tevm actions api spec represented as TypeScript types

**This package has no javascript only typescript types.**

The first implementation of this spec is [@tevm/actions](../actions/) which implements it with Ethereumjs.

## Installation

This package is available in the main `tevm` package

```typescript
import {ScriptHandler} from 'tevm/errors'
```

But can also be installed as a standalone package.

```bash
npm install --save-dev @tevm/errors
```

## License ðŸ“„

<a href="./LICENSE"><img src="https://user-images.githubusercontent.com/35039927/231030761-66f5ce58-a4e9-4695-b1fe-255b1bceac92.png" width="200" /></a>
