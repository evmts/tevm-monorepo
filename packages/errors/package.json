{
  "name": "@tevm/errors",
  "version": "1.0.0-next.28",
  "private": false,
  "description": "A typesafe library for writing forge scripts in typescript",
  "keywords": [
    "solidity",
    "sol",
    "typescript",
    "web3",
    "blockchain"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/evmts/tevm-monorepo.git",
    "directory": "vm/errors"
  },
  "license": "MIT",
  "contributors": [
    "Will Cory <willcory10@gmail.com>"
  ],
  "type": "module",
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": {
        "types": "./types/index.d.ts",
        "default": "./dist/index.js"
      },
      "require": {
        "types": "./dist/index.d.cts",
        "default": "./dist/index.cjs"
      }
    }
  },
  "main": "dist/index.cjs",
  "module": "dist/index.js",
  "types": "types/index.d.ts",
  "files": [
    "dist",
    "src",
    "!src/**/*.spec.ts",
    "types"
  ],
  "scripts": {
    "all": "pnpm i  && bun run build && bun lint && bun format && bun test:run && bun generate:docs && bun lint:deps && bun lint:package",
    "build": "nx run-many --targets=build:dist,build:types --projects=@tevm/errors",
    "build:dist": "tsup",
    "build:types": "tsup --dts-only && tsc --emitDeclarationOnly --declaration",
    "clean": "rm -rf node_modules && rm -rf artifacts && rm -rf dist && rm -rf cache",
    "format": "rome format . --write",
    "format:check": "rome format .",
    "generate:docs": "typedoc",
    "lint": "rome check . --apply-unsafe",
    "lint:check": "rome check . --verbose",
    "lint:deps": "bunx depcheck",
    "lint:package": "bunx publint --strict && attw --pack",
    "package:up": "pnpm up --latest",
    "test": "bun test --watch",
    "test:coverage": "bun test --coverage",
    "test:run": "bun test",
    "typecheck": "tsc --noEmit"
  },
  "devDependencies": {
    "@tevm/tsconfig": "workspace:^",
    "@tevm/tsupconfig": "workspace:^"
  },
  "publishConfig": {
    "access": "public"
  }
}
