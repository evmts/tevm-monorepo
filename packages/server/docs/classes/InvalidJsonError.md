[**@tevm/server**](../README.md) • **Docs**

***

[@tevm/server](../globals.md) / InvalidJsonError

# Class: InvalidJsonError

Represents an error that occurs when parsing JSON fails.

This error is typically encountered when there is an issue with the JSON structure, such as a syntax error or malformed JSON.

## Example

```ts
try {
  const data = parseJsonWithSomeTevmMethod(someString)
} catch (error) {
  if (error instanceof InvalidJsonError) {
    console.error(error.message);
    // Handle the invalid JSON error
  }
}
```

## Param

A human-readable error message.

## Param

Additional parameters for the InvalidJsonError.

## Extends

- `BaseError`

## Constructors

### new InvalidJsonError()

> **new InvalidJsonError**(`message`, `args`?): [`InvalidJsonError`](InvalidJsonError.md)

Constructs an InvalidJsonError.

#### Parameters

• **message**: `string`

Human-readable error message.

• **args?**: [`InvalidJsonErrorParameters`](../interfaces/InvalidJsonErrorParameters.md) = `{}`

Additional parameters for the InvalidJsonError.

#### Returns

[`InvalidJsonError`](InvalidJsonError.md)

#### Overrides

`BaseError.constructor`

#### Defined in

[packages/server/src/errors/InvalidJsonError.js:47](https://github.com/evmts/tevm-monorepo/blob/main/packages/server/src/errors/InvalidJsonError.js#L47)

## Properties

### \_tag

> **\_tag**: `string`

Same as name, used internally.

#### Inherited from

`BaseError._tag`

#### Defined in

packages/errors/types/ethereum/BaseError.d.ts:39

***

### cause

> **cause**: `any`

#### Inherited from

`BaseError.cause`

#### Defined in

packages/errors/types/ethereum/BaseError.d.ts:64

***

### code

> **code**: `number`

Error code, analogous to the code in JSON RPC error.

#### Inherited from

`BaseError.code`

#### Defined in

packages/errors/types/ethereum/BaseError.d.ts:63

***

### details

> **details**: `string`

#### Inherited from

`BaseError.details`

#### Defined in

packages/errors/types/ethereum/BaseError.d.ts:43

***

### docsPath

> **docsPath**: `undefined` \| `string`

Path to the documentation for this error.

#### Inherited from

`BaseError.docsPath`

#### Defined in

packages/errors/types/ethereum/BaseError.d.ts:47

***

### message

> **message**: `string`

Human-readable error message.

#### Inherited from

`BaseError.message`

#### Defined in

node\_modules/.pnpm/typescript@5.5.4/node\_modules/typescript/lib/lib.es5.d.ts:1077

***

### metaMessages

> **metaMessages**: `undefined` \| `string`[]

Additional meta messages for more context.

#### Inherited from

`BaseError.metaMessages`

#### Defined in

packages/errors/types/ethereum/BaseError.d.ts:51

***

### name

> **name**: `string`

The name of the error, used to discriminate errors.

#### Inherited from

`BaseError.name`

#### Defined in

node\_modules/.pnpm/typescript@5.5.4/node\_modules/typescript/lib/lib.es5.d.ts:1076

***

### shortMessage

> **shortMessage**: `string`

#### Inherited from

`BaseError.shortMessage`

#### Defined in

packages/errors/types/ethereum/BaseError.d.ts:55

***

### stack?

> `optional` **stack**: `string`

#### Inherited from

`BaseError.stack`

#### Defined in

node\_modules/.pnpm/typescript@5.5.4/node\_modules/typescript/lib/lib.es5.d.ts:1078

***

### version

> **version**: `string`

#### Inherited from

`BaseError.version`

#### Defined in

packages/errors/types/ethereum/BaseError.d.ts:59

***

### prepareStackTrace()?

> `static` `optional` **prepareStackTrace**: (`err`, `stackTraces`) => `any`

Optional override for formatting stack traces

#### Parameters

• **err**: `Error`

• **stackTraces**: `CallSite`[]

#### Returns

`any`

#### See

https://v8.dev/docs/stack-trace-api#customizing-stack-traces

#### Inherited from

`BaseError.prepareStackTrace`

#### Defined in

node\_modules/.pnpm/@types+node@20.14.8/node\_modules/@types/node/globals.d.ts:28

***

### stackTraceLimit

> `static` **stackTraceLimit**: `number`

#### Inherited from

`BaseError.stackTraceLimit`

#### Defined in

node\_modules/.pnpm/@types+node@20.14.8/node\_modules/@types/node/globals.d.ts:30

## Methods

### walk()

> **walk**(`fn`?): `unknown`

Walks through the error chain.

#### Parameters

• **fn?**: `Function`

A function to execute on each error in the chain.

#### Returns

`unknown`

The first error that matches the function, or the original error.

#### Inherited from

`BaseError.walk`

#### Defined in

packages/errors/types/ethereum/BaseError.d.ts:70

***

### captureStackTrace()

#### captureStackTrace(targetObject, constructorOpt)

> `static` **captureStackTrace**(`targetObject`, `constructorOpt`?): `void`

Create .stack property on a target object

##### Parameters

• **targetObject**: `object`

• **constructorOpt?**: `Function`

##### Returns

`void`

##### Inherited from

`BaseError.captureStackTrace`

##### Defined in

node\_modules/.pnpm/@types+node@20.14.8/node\_modules/@types/node/globals.d.ts:21

#### captureStackTrace(targetObject, constructorOpt)

> `static` **captureStackTrace**(`targetObject`, `constructorOpt`?): `void`

Create .stack property on a target object

##### Parameters

• **targetObject**: `object`

• **constructorOpt?**: `Function`

##### Returns

`void`

##### Inherited from

`BaseError.captureStackTrace`

##### Defined in

node\_modules/.pnpm/@types+node@22.2.0/node\_modules/@types/node/globals.d.ts:22

#### captureStackTrace(targetObject, constructorOpt)

> `static` **captureStackTrace**(`targetObject`, `constructorOpt`?): `void`

Create .stack property on a target object

##### Parameters

• **targetObject**: `object`

• **constructorOpt?**: `Function`

##### Returns

`void`

##### Inherited from

`BaseError.captureStackTrace`

##### Defined in

node\_modules/.pnpm/@types+node@20.12.14/node\_modules/@types/node/globals.d.ts:21

#### captureStackTrace(targetObject, constructorOpt)

> `static` **captureStackTrace**(`targetObject`, `constructorOpt`?): `void`

Create .stack property on a target object

##### Parameters

• **targetObject**: `object`

• **constructorOpt?**: `Function`

##### Returns

`void`

##### Inherited from

`BaseError.captureStackTrace`

##### Defined in

node\_modules/.pnpm/bun-types@1.1.22/node\_modules/bun-types/globals.d.ts:1629
