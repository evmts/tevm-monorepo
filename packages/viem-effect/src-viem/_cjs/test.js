'use strict'
Object.defineProperty(exports, '__esModule', { value: true })
exports.stopImpersonatingAccount =
	exports.setRpcUrl =
	exports.snapshot =
	exports.setStorageAt =
	exports.setNonce =
	exports.setNextBlockTimestamp =
	exports.setNextBlockBaseFeePerGas =
	exports.setMinGasPrice =
	exports.setLoggingEnabled =
	exports.setIntervalMining =
	exports.setCoinbase =
	exports.setCode =
	exports.setBlockTimestampInterval =
	exports.setBlockGasLimit =
	exports.setBalance =
	exports.setAutomine =
	exports.sendUnsignedTransaction =
	exports.revert =
	exports.reset =
	exports.removeBlockTimestampInterval =
	exports.mine =
	exports.inspectTxpool =
	exports.increaseTime =
	exports.impersonateAccount =
	exports.getTxpoolStatus =
	exports.getTxpoolContent =
	exports.getAutomine =
	exports.dropTransaction =
		void 0
const dropTransaction_js_1 = require('./actions/test/dropTransaction.js')
Object.defineProperty(exports, 'dropTransaction', {
	enumerable: true,
	get: function () {
		return dropTransaction_js_1.dropTransaction
	},
})
const getAutomine_js_1 = require('./actions/test/getAutomine.js')
Object.defineProperty(exports, 'getAutomine', {
	enumerable: true,
	get: function () {
		return getAutomine_js_1.getAutomine
	},
})
const getTxpoolContent_js_1 = require('./actions/test/getTxpoolContent.js')
Object.defineProperty(exports, 'getTxpoolContent', {
	enumerable: true,
	get: function () {
		return getTxpoolContent_js_1.getTxpoolContent
	},
})
const getTxpoolStatus_js_1 = require('./actions/test/getTxpoolStatus.js')
Object.defineProperty(exports, 'getTxpoolStatus', {
	enumerable: true,
	get: function () {
		return getTxpoolStatus_js_1.getTxpoolStatus
	},
})
const impersonateAccount_js_1 = require('./actions/test/impersonateAccount.js')
Object.defineProperty(exports, 'impersonateAccount', {
	enumerable: true,
	get: function () {
		return impersonateAccount_js_1.impersonateAccount
	},
})
const increaseTime_js_1 = require('./actions/test/increaseTime.js')
Object.defineProperty(exports, 'increaseTime', {
	enumerable: true,
	get: function () {
		return increaseTime_js_1.increaseTime
	},
})
const inspectTxpool_js_1 = require('./actions/test/inspectTxpool.js')
Object.defineProperty(exports, 'inspectTxpool', {
	enumerable: true,
	get: function () {
		return inspectTxpool_js_1.inspectTxpool
	},
})
const mine_js_1 = require('./actions/test/mine.js')
Object.defineProperty(exports, 'mine', {
	enumerable: true,
	get: function () {
		return mine_js_1.mine
	},
})
const removeBlockTimestampInterval_js_1 = require('./actions/test/removeBlockTimestampInterval.js')
Object.defineProperty(exports, 'removeBlockTimestampInterval', {
	enumerable: true,
	get: function () {
		return removeBlockTimestampInterval_js_1.removeBlockTimestampInterval
	},
})
const reset_js_1 = require('./actions/test/reset.js')
Object.defineProperty(exports, 'reset', {
	enumerable: true,
	get: function () {
		return reset_js_1.reset
	},
})
const revert_js_1 = require('./actions/test/revert.js')
Object.defineProperty(exports, 'revert', {
	enumerable: true,
	get: function () {
		return revert_js_1.revert
	},
})
const sendUnsignedTransaction_js_1 = require('./actions/test/sendUnsignedTransaction.js')
Object.defineProperty(exports, 'sendUnsignedTransaction', {
	enumerable: true,
	get: function () {
		return sendUnsignedTransaction_js_1.sendUnsignedTransaction
	},
})
const setAutomine_js_1 = require('./actions/test/setAutomine.js')
Object.defineProperty(exports, 'setAutomine', {
	enumerable: true,
	get: function () {
		return setAutomine_js_1.setAutomine
	},
})
const setBalance_js_1 = require('./actions/test/setBalance.js')
Object.defineProperty(exports, 'setBalance', {
	enumerable: true,
	get: function () {
		return setBalance_js_1.setBalance
	},
})
const setBlockGasLimit_js_1 = require('./actions/test/setBlockGasLimit.js')
Object.defineProperty(exports, 'setBlockGasLimit', {
	enumerable: true,
	get: function () {
		return setBlockGasLimit_js_1.setBlockGasLimit
	},
})
const setBlockTimestampInterval_js_1 = require('./actions/test/setBlockTimestampInterval.js')
Object.defineProperty(exports, 'setBlockTimestampInterval', {
	enumerable: true,
	get: function () {
		return setBlockTimestampInterval_js_1.setBlockTimestampInterval
	},
})
const setCode_js_1 = require('./actions/test/setCode.js')
Object.defineProperty(exports, 'setCode', {
	enumerable: true,
	get: function () {
		return setCode_js_1.setCode
	},
})
const setCoinbase_js_1 = require('./actions/test/setCoinbase.js')
Object.defineProperty(exports, 'setCoinbase', {
	enumerable: true,
	get: function () {
		return setCoinbase_js_1.setCoinbase
	},
})
const setIntervalMining_js_1 = require('./actions/test/setIntervalMining.js')
Object.defineProperty(exports, 'setIntervalMining', {
	enumerable: true,
	get: function () {
		return setIntervalMining_js_1.setIntervalMining
	},
})
const setLoggingEnabled_js_1 = require('./actions/test/setLoggingEnabled.js')
Object.defineProperty(exports, 'setLoggingEnabled', {
	enumerable: true,
	get: function () {
		return setLoggingEnabled_js_1.setLoggingEnabled
	},
})
const setMinGasPrice_js_1 = require('./actions/test/setMinGasPrice.js')
Object.defineProperty(exports, 'setMinGasPrice', {
	enumerable: true,
	get: function () {
		return setMinGasPrice_js_1.setMinGasPrice
	},
})
const setNextBlockBaseFeePerGas_js_1 = require('./actions/test/setNextBlockBaseFeePerGas.js')
Object.defineProperty(exports, 'setNextBlockBaseFeePerGas', {
	enumerable: true,
	get: function () {
		return setNextBlockBaseFeePerGas_js_1.setNextBlockBaseFeePerGas
	},
})
const setNextBlockTimestamp_js_1 = require('./actions/test/setNextBlockTimestamp.js')
Object.defineProperty(exports, 'setNextBlockTimestamp', {
	enumerable: true,
	get: function () {
		return setNextBlockTimestamp_js_1.setNextBlockTimestamp
	},
})
const setNonce_js_1 = require('./actions/test/setNonce.js')
Object.defineProperty(exports, 'setNonce', {
	enumerable: true,
	get: function () {
		return setNonce_js_1.setNonce
	},
})
const setStorageAt_js_1 = require('./actions/test/setStorageAt.js')
Object.defineProperty(exports, 'setStorageAt', {
	enumerable: true,
	get: function () {
		return setStorageAt_js_1.setStorageAt
	},
})
const snapshot_js_1 = require('./actions/test/snapshot.js')
Object.defineProperty(exports, 'snapshot', {
	enumerable: true,
	get: function () {
		return snapshot_js_1.snapshot
	},
})
const setRpcUrl_js_1 = require('./actions/test/setRpcUrl.js')
Object.defineProperty(exports, 'setRpcUrl', {
	enumerable: true,
	get: function () {
		return setRpcUrl_js_1.setRpcUrl
	},
})
const stopImpersonatingAccount_js_1 = require('./actions/test/stopImpersonatingAccount.js')
Object.defineProperty(exports, 'stopImpersonatingAccount', {
	enumerable: true,
	get: function () {
		return stopImpersonatingAccount_js_1.stopImpersonatingAccount
	},
})
//# sourceMappingURL=test.js.map
