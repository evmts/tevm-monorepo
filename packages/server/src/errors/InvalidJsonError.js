import { BaseError } from '@tevm/errors'

/**
 * Parameters for constructing an {@link InvalidJsonError}.
 * @typedef {Object} InvalidJsonErrorParameters
 * @property {string} [docsBaseUrl] - Base URL for the documentation.
 * @property {string} [docsPath] - Path to the documentation.
 * @property {string} [docsSlug] - Slug for the documentation.
 * @property {string[]} [metaMessages] - Additional meta messages.
 * @property {import('@tevm/errors').BaseError|Error} [cause] - The cause of the error.
 * @property {string} [details] - Details of the error.
 * @property {object} [meta] - Optional object containing additional information about the error.
 */

/**
 * Represents an error that occurs when parsing JSON fails.
 *
 * This error is typically encountered when there is an issue with the JSON structure, such as a syntax error or malformed JSON.
 *
 * @example
 * try {
 *   const data = parseJsonWithSomeTevmMethod(someString)
 * } catch (error) {
 *   if (error instanceof InvalidJsonError) {
 *     console.error(error.message);
 *     // Handle the invalid JSON error
 *   }
 * }
 *
 * @param {string} message - A human-readable error message.
 * @param {InvalidJsonErrorParameters} [args={}] - Additional parameters for the InvalidJsonError.
 * @property {'InvalidJsonError'} _tag - Same as name, used internally.
 * @property {'InvalidJsonError'} name - The name of the error, used to discriminate errors.
 * @property {string} message - Human-readable error message.
 * @property {object} [meta] - Optional object containing additional information about the error.
 * @property {number} code - Error code, analogous to the code in JSON RPC error.
 * @property {string} docsPath - Path to the documentation for this error.
 * @property {string[]} [metaMessages] - Additional meta messages for more context.
 */
export class InvalidJsonError extends BaseError {
	/**
	 * Constructs an InvalidJsonError.
	 *
	 * @param {string} message - Human-readable error message.
	 * @param {InvalidJsonErrorParameters} [args={}] - Additional parameters for the InvalidJsonError.
	 */
	constructor(message, args = {}) {
		super(
			message,
			{
				...args,
				docsBaseUrl: 'https://tevm.sh',
				docsPath: '/reference/tevm/errors/classes/invalidjsonerror/',
			},
			'InvalidJsonError',
			-32700,
		)
	}
}
