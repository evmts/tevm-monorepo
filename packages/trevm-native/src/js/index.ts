import { createLogger } from '@tevm/logger';
import { InternalError } from '@tevm/errors';
import type { StateManager } from '@tevm/state';
import type { Common } from '@tevm/common';
import type { Chain } from '@tevm/blockchain';

// Import native module
import { init_trevm, TrevmWrapper } from '../../dist/index.node';

export interface TrevmEvmOptions {
  common: Common;
  stateManager: StateManager;
  blockchain: Chain;
  customPrecompiles?: any[];
  profiler?: boolean;
  allowUnlimitedContractSize?: boolean;
  loggingLevel?: string;
}

export interface TrevmRunCallOptions {
  caller?: string;
  to?: string;
  data?: string;
  value?: bigint;
  gasLimit?: bigint;
  skipBalance?: boolean;
}

export interface TrevmRunCallResult {
  returnValue: string;
  executionGasUsed: string;
  error?: string;
}

export interface TrevmEvm {
  runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult>;
}

/**
 * Creates a trevm-based EVM instance
 * @param options EVM options
 * @returns A trevm EVM instance
 */
export const createTrevm = async ({
  common,
  stateManager,
  blockchain,
  customPrecompiles,
  profiler,
  allowUnlimitedContractSize,
  loggingLevel,
}: TrevmEvmOptions): Promise<TrevmEvm> => {
  const logger = createLogger({
    name: '@tevm/trevm',
    level: loggingLevel ?? 'warn',
  });
  
  logger.debug({
    allowUnlimitedContractSize,
    profiler,
    customPrecompiles: customPrecompiles?.map(c => c.address.toString()),
  });

  try {
    // Initialize the native trevm module with callbacks to the StateManager
    const nativeTrevm = await init_trevm({
      chainId: common.chainId(),
      allowUnlimitedContractSize: allowUnlimitedContractSize ?? false,
      stateCallbacks: {
        getAccount: async (address: string) => {
          try {
            const account = await stateManager.getAccount(address);
            if (!account) return null;
            
            return {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? `0x${Buffer.from(account.codeHash).toString('hex')}` : undefined,
              code: account.code ? `0x${Buffer.from(account.code).toString('hex')}` : undefined,
            };
          } catch (error) {
            logger.error(`Error in getAccount for ${address}:`, error);
            return null;
          }
        },
        putAccount: async (address: string, account: any) => {
          try {
            await stateManager.putAccount(address, {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? Buffer.from(account.codeHash.slice(2), 'hex') : undefined,
              code: account.code ? Buffer.from(account.code.slice(2), 'hex') : undefined,
            });
            return true;
          } catch (error) {
            logger.error(`Error in putAccount for ${address}:`, error);
            return false;
          }
        },
        getStorage: async (address: string, key: string) => {
          try {
            const value = await stateManager.getContractStorage(address, key);
            return value ? `0x${Buffer.from(value).toString('hex')}` : '0x0';
          } catch (error) {
            logger.error(`Error in getStorage for ${address} key ${key}:`, error);
            return '0x0';
          }
        },
        putStorage: async (address: string, key: string, value: string) => {
          try {
            await stateManager.putContractStorage(
              address,
              key,
              Buffer.from(value.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putStorage for ${address} key ${key}:`, error);
            return false;
          }
        },
        getCode: async (address: string) => {
          try {
            const code = await stateManager.getContractCode(address);
            return code ? `0x${Buffer.from(code).toString('hex')}` : '0x';
          } catch (error) {
            logger.error(`Error in getCode for ${address}:`, error);
            return '0x';
          }
        },
        putCode: async (address: string, code: string) => {
          try {
            await stateManager.putContractCode(
              address,
              Buffer.from(code.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putCode for ${address}:`, error);
            return false;
          }
        },
        getBlockHash: async (blockNumber: number) => {
          try {
            const block = await blockchain.getBlock(BigInt(blockNumber));
            return block ? `0x${Buffer.from(block.header.hash()).toString('hex')}` : null;
          } catch (error) {
            logger.error(`Error in getBlockHash for block ${blockNumber}:`, error);
            return null;
          }
        }
      }
    });

    // Return a JavaScript wrapper around the native trevm that matches the Evm interface
    return {
      async runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult> {
        try {
          return await nativeTrevm.run_call({
            caller: opts.caller,
            to: opts.to,
            data: opts.data,
            value: opts.value,
            gas_limit: opts.gasLimit,
            skip_balance: opts.skipBalance,
          });
        } catch (error) {
          logger.error('Error in runCall:', error);
          throw new InternalError(`Trevm execution error: ${error}`);
        }
      }
    };
  } catch (error) {
    logger.error('Error initializing trevm:', error);
    throw new InternalError(`Failed to initialize trevm: ${error}`);
  }
};