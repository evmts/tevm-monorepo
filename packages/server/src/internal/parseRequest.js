import { InvalidRequestError } from '@tevm/errors'
import { z } from 'zod'
import { InvalidJsonError } from '../errors/InvalidJsonError.js'

const zJsonRpcRequest = z.object({
	jsonrpc: z.literal('2.0').optional(),
	id: z.union([z.string(), z.number(), z.null()]).optional(),
	method: z.string(),
	params: z.unknown().optional(),
})

const zBulkRequest = z.array(zJsonRpcRequest)

/**
 * Parses a request body into a JSON-RPC request object.
 * Returns error if any
 * @param {string} body
 * @throws {never} returns errors as values
 */
export const parseRequest = (body) => {
	/**
	 * @type {unknown}
	 */
	let raw
	try {
		raw = JSON.parse(body)
	} catch (e) {
		const err = /** @type {Error} */ (e)
		return new InvalidJsonError(err.message, { cause: err })
	}
	const parsedRequest = Array.isArray(raw) ? zBulkRequest.safeParse(raw) : zJsonRpcRequest.safeParse(raw)
	if (!parsedRequest.success) {
		return new InvalidRequestError(JSON.stringify(parsedRequest.error.format()))
	}
	return parsedRequest.data
}
