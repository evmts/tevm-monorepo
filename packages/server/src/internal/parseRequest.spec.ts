import { describe, expect, it } from 'bun:test'
import { InvalidRequestError } from '@tevm/errors'
import { InvalidJsonError } from '../errors/InvalidJsonError.js'
import { parseRequest } from './parseRequest.js'

describe('parseRequest', () => {
	const validJsonRpcRequest = {
		jsonrpc: '2.0',
		method: 'tevm_call',
		params: [],
		id: 1,
	} as const

	const invalidJsonRpcRequest = {
		jsonrpc: '2.0',
		method: 'tevm_call',
		params: 'invalid_params', // Should be an array
		id: 1,
	} as const

	it('should parse a valid single JSON-RPC request', () => {
		const body = JSON.stringify(validJsonRpcRequest)
		const result = parseRequest(body)
		expect(result).toEqual(validJsonRpcRequest)
	})

	it('should parse a valid bulk JSON-RPC request', () => {
		const body = JSON.stringify([validJsonRpcRequest, validJsonRpcRequest])
		const result = parseRequest(body)
		expect(result).toEqual([validJsonRpcRequest, validJsonRpcRequest])
	})

	it('should return InvalidJsonError for invalid JSON', () => {
		const body = '{ "jsonrpc": "2.0", "method": "tevm_call", "params": [ ] ' // Malformed JSON
		const result = parseRequest(body)
		expect(result).toBeInstanceOf(InvalidJsonError)
		expect((result as InvalidJsonError).message).toMatchSnapshot()
	})

	it('should return InvalidRequestError for invalid JSON-RPC request', () => {
		const body = JSON.stringify(invalidJsonRpcRequest)
		const result = parseRequest(body)
		expect(result).toBeInstanceOf(InvalidRequestError)
		expect((result as InvalidRequestError).message).toMatchSnapshot()
	})

	it('should return InvalidRequestError for invalid bulk JSON-RPC request', () => {
		const body = JSON.stringify([validJsonRpcRequest, invalidJsonRpcRequest])
		const result = parseRequest(body)
		expect(result).toBeInstanceOf(InvalidRequestError)
		expect((result as InvalidRequestError).message).toMatchSnapshot()
	})
})
