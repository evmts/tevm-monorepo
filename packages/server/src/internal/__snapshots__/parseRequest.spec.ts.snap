// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseRequest > should return InvalidJsonError for invalid JSON 1`] = `
"Expected ',' or '}' after property value in JSON at position 57 (line 1 column 58)

Docs: https://tevm.sh/reference/tevm/errors/classes/invalidjsonerror/
Details: Expected ',' or '}' after property value in JSON at position 57 (line 1 column 58)
Version: 1.1.0.next-73"
`;

exports[`parseRequest > should return InvalidRequestError for invalid JSON-RPC request 1`] = `
"{"_errors":[],"method":{"_errors":["Required"]}}

Docs: https://tevm.sh/reference/tevm/errors/classes/invalidrequesterror/
Version: 1.1.0.next-73"
`;

exports[`parseRequest > should return InvalidRequestError for invalid bulk JSON-RPC request 1`] = `
"{"1":{"_errors":[],"method":{"_errors":["Required"]}},"_errors":[]}

Docs: https://tevm.sh/reference/tevm/errors/classes/invalidrequesterror/
Version: 1.1.0.next-73"
`;
